subsys/bluetooth/shell/gatt.c

```diff
@@ -85,19 +85,22 @@ static void reset_write_stats(void)
 	memset(&write_stats, 0, sizeof(write_stats));
 }
 
+/* This variable is write-locked when `(exchange_params.func != NULL)`.
+ * Must be zero-initialized when unlocked.
+ */
+static struct bt_gatt_exchange_params exchange_params;
+
 static void exchange_func(struct bt_conn *conn, uint8_t err,
 			  struct bt_gatt_exchange_params *params)
 {
 	shell_print(ctx_shell, "Exchange %s", err == 0U ? "successful" :
 		    "failed");
 
-	(void)memset(params, 0, sizeof(*params));
```

Comment:
Wouldn't it be sufficient to remove this `memset` so that `func` would be non-null?

Reply:
No. `params` is conceptually locked by `bt_gatt_exchange_mtu`. If the user of the shell were to run a `gatt exchange-params`  before the previous is completed, we would have a collision. Imagine if we add a `bt select-conn <index>`. Then the user may be tempted to paste a series of commands. `gatt exchange-params` `bt select-conn 1` `gatt exchange-params`.

Reply:
I'm afraid I don't follow. `bt_gatt_exchange_params` is `conn` independent. It keeps the `func` callback pointer to be called once operation is complete.

```
struct bt_gatt_exchange_params {
	/** Response callback */
	void (*func)(struct bt_conn *conn, uint8_t err,
		     struct bt_gatt_exchange_params *params);
};
```

Even if the `default_conn` would change when the operation is in progress, there would be no collision here IMO.

Reply:
I agree with Mariusz here, there is no need to over-engineer this. Removing memset should be enough.
What is more, MTU procedure is not "global", ie you can execute it concurrently on different connection.

Reply:
> I agree with Mariusz here, there is no need to over-engineer this. Removing memset should be enough. What is more, MTU procedure is not "global", ie you can execute it concurrently on different connection.

Please have a think. This is precisely my point. We cannot use the same `bt_gatt_exchange_params` concurrently.

Reply:
Why are you saying it's over engineered? Will it be a maintenance burden? Is it difficult to follow the code? I gather that it's difficult to see that this is needed for correctness. Is there something I can do to convince you?

The pattern in the code is basically a dynamic allocation from a pool of size one. The allocated object is stipulated to be zero-initialized for convenience. It's not very complicated/over-engineered. It must be like that because `bt_gatt_exchange_mtu` requires exclusive use of the `bt_gatt_exchange_params` object.

Reply:
I should say: The reason I think `bt_gatt_exchange_mtu` needs exclusive use of the `bt_gatt_exchange_params` is that it's not passed in as const. There is a pattern in our stack that we extend params-structs when needed. Clearly, the reason it's not const, is that we have left ourselves room to add internal bookkeeping like "node", etc.

The above reasoning was enough to make me believe that using a `bt_gatt_exchange_params` concurrently will break. But that is not true with the current implementation. So I will give you that there will not be a collision with the current implementation, the reasoning about correctness still stands. The alternative is to make `bt_gatt_exchange_mtu` take a const.

Reply:
ok, I see your point that this could cause issues if params are extended with "per call" storage,